luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,156 @@
+local trackingTable = {}
+local tagUpdateFrameMax = 1 -- Can be bigger than 1 if you want game freezes.
+
+local statuses = {
+	outOfRange = 0,
+	inRange = 1,
+}
+
+local mobileUnits = {}
+local constructionTurretsDefs = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+
+	if unitDef.canMove then
+		mobileUnits[unitDefID] = true
+	end
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+	gadgetHandler:RegisterAllowCommand(CMD.STOP)
+end
+
+function gadget:GameFrame(frame)
+	local tagUpdates = 0
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if commandData.cmdTag == -1 then
+						if tagUpdates >= tagUpdateFrameMax then break end
+
+						if not cmdCache[nanoID] then
+							cmdCache[nanoID] = spGetUnitCommands(nanoID, -1)
+							tagUpdates = tagUpdates +1
+						end
+
+						for _, cmd in ipairs(cmdCache[nanoID]) do
+							trackingTable[nanoID][cmd.params[1]].cmdTag = cmd.tag
+						end
+					end
+
+					local distance = spGetUnitSeparation(nanoID, targetID, constructionTurretsDefs[nanoDefID].buildRange3D, false)
+
+					if distance < maxDistance then
+						if commandData.status == statuses.outOfRange then
+							commandData.status = statuses.inRange
+						end
+					else
+						-- Do not blindly remove orders on out-of-range mobile units, as the player may have pre-emptively issued an order before the unit was in range
+						-- Instead, only remove the order once the unit has come into range then left again
+						if commandData.status == statuses.inRange then
+							spGiveOrderToUnit(nanoID, CMD.REMOVE, commandData.cmdTag, 0)
+							targets[targetID] = nil
+						end
+					end
+				else
+					targets[targetID] = nil
+				end
+			end
+		end
+	end
+end
+
+function gadget:AllowCommand(unitID, unitDefID, teamID, cmdID,	cmdParams, cmdOptions, cmdTag, synced, fromLua)
+	local constructionTurretDef = constructionTurretsDefs[unitDefID]
+	if not constructionTurretDef then return true end
+
+	if cmdID == CMD_STOP then
+		trackingTable[unitID] = nil
+	end
+
+	-- Rather than trying to handle area commands, let them fallthrough, then catch the resulting list of single-target commands
+	if #cmdParams ~= 1 then return true end
+
+	local distance = math.huge
+	local targetId = cmdParams[1]
```

Comment:
The `gMaxUnits` check is a bit magic. Not necessarily in the scope for this PR (imo fine as-is) but perhaps there could be a helper like below until engine sorts this mess out:
```lua
function Spring.Utilities.ParseUnitOrFeatureID(targetID) -- isUnit, actualID
	if targetID < Game.maxUnits then
		return true, targetID
	else
		return false, targetID - Game.maxUnits
	end
end
```

Reply:
There is already a hand full of operations with Game.maxUnits, I just would leave it like that until a universal fix happens so all instances can be searched more easily and refactored, instead of there being two ways that people start to do things.